prov/rxm/src/rxm_eq.c

```diff
@@ -64,6 +64,55 @@ static struct fi_ops rxm_eq_fi_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
+ssize_t rxm_eq_write(struct fid_eq *eq_fid, uint32_t event,
+		     const void *buf, size_t len, uint64_t flags)
+{
+	struct fi_eq_entry entry;
+	const struct fi_eq_entry *in_entry = buf;
+	struct rxm_mc *mc;
+
+	if (event != FI_JOIN_COMPLETE) {
+		return ofi_eq_write(eq_fid, event, buf, len, flags);
+	}
+
+	mc = in_entry->context;
+	if (in_entry->fid == &mc->util_coll_mc_fid->fid) {
+		/* cleanup after partially executed fi_join() */
+		if (mc->util_coll_join_completed == -1) {
+			mc->util_coll_join_completed = 1;
+			fi_close(&mc->mc_fid.fid);
```

Comment:
This is racy with ep close trying to free the underlying joins.

Reply:
@shefty shall we add another lock here or is it OK to use just usleep() and cyclically poll for mc.state?

Reply:
A sleep call is almost never acceptable

Reply:
Done.